resource_processor/vmss_porter/runner.py

```diff
@@ -82,40 +85,64 @@ async def filter_parameters_not_needed_by_porter(msg_body, env_vars):
     if stdout:
         result_stdout = stdout.decode()
         porter_explain_parameters = json.loads(result_stdout)["parameters"]
-        items = [item["name"] for item in porter_explain_parameters]
-        porter_keys = set(items).intersection(set(parameters.keys()))
-        return porter_keys
+        porter_parameter_keys = [item["name"] for item in porter_explain_parameters]
+        return porter_parameter_keys
     if stderr:
         result_stderr = stderr.decode()
         logger_adapter.info('[stderr]')
         for string in result_stderr.split('\n'):
             logger_adapter.info(str(string))
 
-    return parameters.keys()
+
+def get_installation_id(msg_body):
+    # this will be used to identify each bundle install within the porter state store.
+    return msg_body['id']
 
 
 async def build_porter_command(msg_body, env_vars):
+    porter_parameter_keys = await get_porter_parameter_keys(msg_body, env_vars)
     porter_parameters = ""
 
-    porter_keys = await filter_parameters_not_needed_by_porter(msg_body, env_vars)
-    for parameter in porter_keys:
-        porter_parameters = porter_parameters + f" --param {parameter}=\"{msg_body['parameters'][parameter]}\""
+    for parameter_name in porter_parameter_keys:
+        # first try to find the param in the msg body parameters collection
+        parameter_value = msg_body['parameters'].get(parameter_name)
+
+        # if not found, try to get it from the env_vars (e.g. terraform state ones)
+        if parameter_value is None:
+            parameter_value = env_vars.get(parameter_name)
 
-    installation_id = msg_body['parameters']['tre_id'] + "-" + msg_body['parameters']['workspace_id']
+        # if not found, try to get it from the msg body root (e.g. id of the resource)
+        if parameter_value is None:
+            parameter_value = msg_body.get(parameter_name)
 
-    porter_parameters = porter_parameters + f" --param tfstate_container_name={env_vars['tfstate_container_name']}"
-    porter_parameters = porter_parameters + f" --param tfstate_resource_group_name={env_vars['tfstate_resource_group_name']}"
-    porter_parameters = porter_parameters + f" --param tfstate_storage_account_name={env_vars['tfstate_storage_account_name']}"
-    porter_parameters = porter_parameters + f" --param arm_use_msi={env_vars['arm_use_msi']}"
+        # if still not found, might be a special case
+        if parameter_value is None:
```

Comment:
You know that the parameter_value is None so why not move this whole block at the start and break from the loop if parameter_name in ["mgmt_acr_name", "mgmt_resource_group_name"] as those are special. Infact could refactor method out to special_param_handling to make the intention clear.

Reply:
I wanted to give an option for the api to supply those values in the msg params, which is why I left it at the end.